tsdb/tsi1/report.go

```diff
@@ -0,0 +1,374 @@
+	}
+
+	c.short = append(c.short, uint32(x)) // Series IDs never get beyond 2^32
+
+	// Cheaper to store in bitmap.
+	if len(c.short) > useBitmapN {
+		c.set = tsdb.NewSeriesIDSet()
+		for i := 0; i < len(c.short); i++ {
+			c.set.AddNoLock(tsdb.NewSeriesID(uint64(c.short[i])))
+		}
+		c.short = nil
+		return
+	}
+}
+
+func (c *cardinality) cardinality() int64 {
+	if c == nil || (c.short == nil && c.set == nil) {
+		return 0
+	}
+
+	if c.short != nil {
+		return int64(len(c.short))
+	}
+	return int64(c.set.Cardinality())
+}
+
+func (report *ReportCommand) calculateCardinalities() error {
+	idx := report.indexFile
+	itr, err := idx.MeasurementIterator()
+	if err != nil {
+		return err
+	} else if itr == nil {
+		return nil
+	}
+	defer itr.Close()
+
+OUTER:
+	for {
+		name, err := itr.Next()
+		if err != nil {
+			return err
+		} else if name == nil {
+			break OUTER
+		}
+
+		// decode org and bucket from measurement name
+		var a [16]byte
+		copy(a[:], name[:16])
+		org, bucket := tsdb.DecodeName(a)
+
+		if report.OrgID != nil && *report.OrgID != org {
+			continue
+		} else if report.BucketID != nil && *report.BucketID != bucket {
+			continue
+		}
+
+		sitr, err := idx.MeasurementSeriesIDIterator(name)
+		if err != nil {
+			return err
+		} else if sitr == nil {
+			continue
+		}
+
+		var bucketCard *cardinality
+
+		// initialize map of bucket to measurements
+		if _, ok := report.byBucketMeasurement[bucket]; !ok {
+			report.byBucketMeasurement[bucket] = make(map[string]*cardinality)
+		}
+
+		if _, ok := report.byOrgBucket[org]; !ok {
+			report.byOrgBucket[org] = make(map[influxdb.ID]*cardinality)
+		}
+
+		// initialize total cardinality tracking struct for this bucket
+		if c, ok := report.byOrgBucket[org][bucket]; !ok {
+			bucketCard = &cardinality{name: []byte(bucket.String())}
+			report.byOrgBucket[org][bucket] = bucketCard
+		} else {
+			bucketCard = c
+		}
+
+		var e tsdb.SeriesIDElem
+		for e, err = sitr.Next(); err == nil && e.SeriesID.ID != 0; e, err = sitr.Next() {
+			id := e.SeriesID.ID
+
+			if id > math.MaxUint32 {
+				panic(fmt.Sprintf("series ID is too large: %d (max %d). Corrupted series file?", e.SeriesID, uint32(math.MaxUint32)))
+			}
+
+			// add cardinality to bucket
+			bucketCard.add(id)
+
+			// retrieve tags associated with series id so we can get
+			// associated measurement
+			_, tags := report.sfile.Series(e.SeriesID)
+			if len(tags) == 0 {
+				panic(fmt.Sprintf("empty series key"))
+			}
+
+			// measurement name should be first tag
+			mName := string(tags[0].Value)
+
+			// update measurement-level cardinality if tracking by measurement
+			if report.ByMeasurement {
+				var mCard *cardinality
+				if cardForM, ok := report.byBucketMeasurement[bucket][mName]; !ok {
+					mCard = &cardinality{name: []byte(mName)}
+					report.byBucketMeasurement[bucket][mName] = mCard
+				} else {
+					mCard = cardForM
+				}
+
+				mCard.add(id)
+			}
+		}
+
+		sitr.Close()
+
+		if err != nil {
+			return err
+		}
```

Comment:
The `sitr.Close()` & `err` check aren't needed here if they're handled with a defer as mentioned above and the `sitr.Next()` handles the check.